Add RPGLE language

barrettotte opened the pull request:
## Description
Add support for IBM RPGLE ([IBM documentation](https://www.ibm.com/support/knowledgecenter/ssw_ibm_i_74/rzasd/rzasdmain.htm)) to help make the language more discoverable in the open source world.
This language is still widely used in insurance and financial institutions and it is a shame more people don't know about it.

## Checklist:
- [x] **I am adding a new language.**
  - [ ] The extension of the new language is used in hundreds of repositories on GitHub.com 
    - I am not quite sure how to get a unique count of repos using advanced search, is there a different magic url for it?
    - Search results for each extension: 
      - (rpgle - 40K files) https://github.com/search?q=extension%3Arpgle+extension%3ARPGLE+NOT+nothack&type=Code
      - (sqlrpgle 4K files) https://github.com/search?q=extension%3Asqlrpgle+extension%3ASQLRPGLE+NOT+nothack&type=Code
  - [x] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - [sms.sqlrpgle](https://github.com/barrettotte/RPGLE-Twilio/blob/master/src/sms.sqlrpgle)
      - [sms_h.rpgle](https://github.com/barrettotte/RPGLE-Twilio/blob/master/src/sms_h.rpgle)
      - [anilist.sqlrpgle](https://github.com/barrettotte/Anilist-IBMi/blob/master/src/anilist.sqlrpgle)
      - [bfint.sqlrpgle](https://github.com/barrettotte/RPGLE-BF/blob/master/src/bfint.sqlrpgle)
      - [ifsread.rpgle](https://github.com/barrettotte/RPGLE-BF/blob/master/src/ifsread.rpgle)
    - Sample license(s):
      - [RPGLE-Twilio (MIT)](https://github.com/barrettotte/RPGLE-Twilio/blob/master/LICENSE)
      - [Anilist-IBMi (MIT)](https://github.com/barrettotte/Anilist-IBMi/blob/master/LICENSE)
      - [RPGLE-BF (MIT)](https://github.com/barrettotte/RPGLE-BF/blob/master/LICENSE)
      - All other sample files included were written today from scratch to finalize my grammars.
  - [ ] I have included a change to the heuristics to distinguish my language from others using the same extension.
  - [x] I have included a syntax highlighting grammar: https://github-lightshow.herokuapp.com/?utf8=%E2%9C%93&scope=from-url&grammar_format=auto&grammar_url=https%3A%2F%2Fraw.githubusercontent.com%2Fbarrettotte%2Fvscode-ibmi-languages%2Fmaster%2Fsyntaxes%2Frpgle.tmLanguage.json&grammar_text=&code_source=from-url&code_url=https%3A%2F%2Fraw.githubusercontent.com%2Fbarrettotte%2FAnilist-IBMi%2Fmaster%2Fsrc%2Fanilist.sqlrpgle&code=

lildude commented:
Now this is surprising... the vast number of uses of these extensions are isolated to a surprisingly few number of repos:

- `.sqlrpgle`

Total files found: 127
Unique public user/repos: 43
Unique owners: 33

- `.rpgle`

Total files found: 25540
Unique public user/repos: 103
Unique owners: 96

- `.RPGLE`

Total files found: 15083
Unique public user/repos: 48
Unique owners: 42

- `.SQLRPGLE`

Total files found: 4480
Unique public user/repos: 53
Unique owners: 47

Unfortunately none in sufficient numbers for inclusion just yet as we use 200 unique user/repo as the threshold for the "hundreds of repositories" requirement for each extension.

barrettotte commented:
@lildude Ah darn, so each extension listed (including the uppercase variants) would have to have 200 unique users/repos *each* to be included? Despite all extensions using the same grammar file and classified as the same language? 

I was actually under the impression there were more RPGLE repos floating around out there than that, but today I learned something at least lol. Thanks for looking!

Nixinova commented:
Linguist is case insensitive, right? So rpgle has 151 repos and sqlrpgle 96. rpgle is not far off then.

barrettotte commented:
@Nixinova Ohhh yeah good point I didn't realize it was case insensitive. I must not have let enough time pass for the linguist to run when testing in a temp repo yesterday

lildude commented:
Sorry, yes, I meant to say Linguist is not case-sensitive when I made my comment. Thanks pointing that out @Nixinova 🙇

charles-wilt commented:
I assume that if/when this gets included, it will also be available to GitHub Enterprise users.  That being the case, I don't suppose there's a way to see counts from GHE repos?  My company has a number of repos in GHE with RPGLE code that we'd love to have recognized.   Our main repo has 1600 sqlrpgle and 4000 rpgle files.

NattyNarwhal commented:
`*.rpg` (as well as the 36/38 variants) is probably another valid extension; but that probably implies fixed format which I don't know if this PR covers.

It doesn't really help that most RPG is internal rather than freely available; I'd expect more GH Enterprise users.

I'd also mention there are other i-specific syntax that might be worth covering later (i.e menus, linker scripts, etc.); the system's scripting language often has `*.clp`, which gets it mistagged as CLIPS.

lildude commented:
> I assume that if/when this gets included, it will also be available to GitHub Enterprise users. 

Yes, but only the most recent version of GitHub Enterprise Server (GHES), currently 3.0.x, gets Linguist updates.

> That being the case, I don't suppose there's a way to see counts from GHE repos?

Nope. GHES is deliberately designed to be completely independent of GitHub.com so no information on GHES is searchable on GitHub.com, unless your instance has GitHub Connect with search enabled, but even then, the search is limited to members of your organisation and not the general public.

bencz commented:
Oh Yeah!! This is will be great!!!
Up!

lildude commented on samples/RPGLE/hello_fixed.rpgle:
```diff
@@ -0,0 +1,2 @@
+     C     'Hello World' DSPLY
+     C                   SETON                                        LR
```
Please remove this. "Hello world" are not representative of real-world use.

lildude commented on samples/RPGLE/hello_free.rpgle:
```diff
@@ -0,0 +1,12 @@
+
+    *inlr = *on;
+    return;
+  end-proc;
```
Please remove this. "Hello world" are not representative of real-world use.

lildude reviewed (changes requested):
Please can you address the conflicts and remove the hello-world samples.

Once this is good to go, I'll approve this based on the temporary requirements detailed in https://github.com/github/linguist/issues/5756

barrettotte commented:
@lildude Thanks for posting on this PR with that update. I believe I have fixed the merge conflicts and removed the hello world samples as requested. Please let me know if I can help any other way.

lildude commented:
🤦 I have no idea why the indentation has gone wonky on the `grammars.yml`. I'll sort it now.

lildude commented:
> 🤦 I have no idea why the indentation has gone wonky on the `grammars.yml`. I'll sort it now.

I do now. https://github.com/github/linguist/pull/5240/commits/fef64e937a8b4bf4e557c79142ef64c2df76ebfb is what intro'd it.

barrettotte commented:
> > 🤦 I have no idea why the indentation has gone wonky on the `grammars.yml`. I'll sort it now.
> 
> I do now. [fef64e9](https://github.com/github/linguist/commit/fef64e937a8b4bf4e557c79142ef64c2df76ebfb) is what intro'd it.

Oh man...sorry I really should have saw the line count on that file and knew something was wrong. Any suggestion on what I should do to fix it? I'm assuming this means I didn't re-sync my fork properly?

lildude commented:
> Any suggestion on what I should do to fix it? I'm assuming this means I didn't re-sync my fork properly?

Fixed. I think you may have the `grammars.yml` file open in your editor and it automatically indented things.
